test/support/protocol_utils.rb

```diff
@@ -0,0 +1,343 @@
+      assert_eval_result 'watch', expression, expected, frame_idx
+    end
+
+    # Not API
+
+    def req_disconnect
+      send_request 'disconnect',
+                    restart: false,
+                    terminateDebuggee: false
+      assert_disconnect_result
+    end
+
+    def assert_disconnect_result
+      res = find_crt_response
+      assert_dap_res :DisconnectResponse, res
+      @reader_thread.raise Detach
+      @sock.close
+    end
+
+    def attach_to_rdbg
+      @sock = Socket.unix @remote_info.sock_path
+      @seq = 1
+      @reader_thread = Thread.new do
+        while res = recv_response
+          @queue.push res
+        end
+      rescue Detach
+      end
+      sleep 0.001 while @reader_thread.status != 'sleep'
+      @reader_thread.run
+      INITIALIZE_DAP_MSGS.each{|msg| send **msg}
+    end
+    
+    def assert_eval_result context, expression, expected, frame_idx
+      send_request 'stackTrace',
+                    threadId: 1,
+                    startFrame: 0,
+                    levels: 20
+      res = find_crt_response
+      f_id = res.dig(:body, :stackFrames, frame_idx, :id)
+      send_request 'evaluate',
+                    expression: expression,
+                    frameId: f_id,
+                    context: context
+      res = find_crt_response
+      assert_dap_res :EvaluateResponse, res
+
+      failure_msg = FailureMessage.new{create_protocol_message "result:\n#{JSON.pretty_generate res}"}
+      if expected.is_a? String
+        expected_val = expected
+      else
+        expected_val = expected.inspect
+      end
+      result_val = res.dig(:body, :result)
+      assert_equal expected_val, result_val, failure_msg
+
+      expected_type = expected.class.inspect
+      result_type = res.dig(:body, :type)
+      assert_equal expected_type, result_type, failure_msg
+    end
+
+    def send_request command, **kw
+      send type: 'request',
+            command: command,
+            arguments: kw
+    end
+
+    def send **kw
+      kw[:seq] = @seq += 1
+      str = JSON.dump(kw)
+      @sock.write "Content-Length: #{str.bytesize}\r\n\r\n#{str}"
+      @backlog << "V>D #{str}"
+    end
+
+    TIMEOUT_SEC = (ENV['RUBY_DEBUG_TIMEOUT_SEC'] || 10).to_i
+
+    def assert_dap_res expected_def, result_res
+      properties = @protocol.dig(:definitions, expected_def, :allOf, 1)
+      failure_msg = FailureMessage.new{create_protocol_message "expected:\n#{JSON.pretty_generate properties}\n\nresult:\n#{JSON.pretty_generate result_res}"}
+      patterns = DAP_ResponsePattern.new.get properties
+      patterns.each{|keys, expected|
+        result = result_res.dig(*keys)
+        assert_kind_of expected, result, failure_msg
+      }
+    end
+
+    def fetch_dap_json
+      return if File.exist? DAP_JSON_PATH
+
+      json = Net::HTTP.get(URI.parse('https://microsoft.github.io/debug-adapter-protocol/debugAdapterProtocol.json'))
+      File.write DAP_JSON_PATH, json
```

Comment:
Shouldn't we store this file as a source of truth? From this method, it looks like each developer and the CI could have different versions of the protocol if it's changed?
If we want to keep up with the latest protocol, we can have a CI action comparing the project's version and the latest version.

Reply:
> If we want to keep up with the latest protocol, we can have a CI action comparing the project's version and the latest version.

Yes, I want to keep up with the latest version, but I don't know how to do it without too much requesting.